### PR TITLE
Integration: 3DEqualizer integration

### PR DESCRIPTION
**Pull request by antirotor**
## Changelog Description
This PR is adding basic integration for [3DEqualizer4 from Science-D-Vision](https://www.3dequalizer.com/). Integration includes:
- Workfiles
- Loading plates (cameras)
- Publishing scripts to Maya and Nuke
- Publishing of lens data

## Additional info
This integration is installing PySide2 into 3DEqualizer environment as it doesn't ship with Qt support. This comes with some price - to make Qt UI work with  3DEqualizer, `processEvent()` is periodically called. This is not optimal and it might create some issues, like 3Dequalizer crashing or UI lags.

For extraction, it is using 3de4 native scripts, but since they are depending on some UI, there are few hacks around it but I think it is better than rewritting the whole export logic to our plugins.

### CODE REVIEW

File: openpype/hosts/equalizer/__init__.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+from .addon import (
+    EqualizerAddon,
+    EQUALIZER_HOST_DIR,
+)
```
Comment (BigRoy) [inline openpype/hosts/equalizer/__init__.py:4]: This I believe is missing:

```suggestion
)


__all__ = [
    "EqualizerAddon",
    "EQUALIZER_HOST_DIR",
]

```

File: openpype/hosts/equalizer/addon.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    name = "equalizer"
+    host_name = "equalizer"
+
+    def initialize(self, module_settings):
+        self.enabled = True
+
+    def add_implementation_envs(self, env, _app):
+        # 3dequalizer utilize PYTHON_CUSTOM_SCRIPTS_3DE4 for custom scripts
+        # and TDE4_ROOT for its root directory. Note that the custom script
+        # paths are semicolon separated even on Windows.
```
Comment (BigRoy) [inline openpype/hosts/equalizer/addon.py:17]: Default on Windows is actually semi-colon separated `;` - did you mean that on windows it's **colon** `:` separated?
```suggestion
        # paths are colon `:` separated even on Windows.
```

File: openpype/hosts/equalizer/api/host.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+
+class EqualizerHost(HostBase, IWorkfileHost, ILoadHost, IPublishHost):
+    name = "equalizer"
+    _instance = None
+
+    def __new__(cls):
+        if not hasattr(cls, "_instance") or not cls._instance:
+            cls._instance = super(EqualizerHost, cls).__new__(cls)
+        return cls._instance
```
Comment (BigRoy) [inline openpype/hosts/equalizer/api/host.py:43]: Why is this needed? I don't see other hosts implementations doing this?

If there are certain reasons why this is required for 3DEqualizer it might be good to add a comment as to WHY that's the case.

File: openpype/hosts/equalizer/api/host.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        return not bool(tde4.isProjectUpToDate())
+
+    def get_workfile_extensions(self):
+        return [".3de"]
+
+    def save_workfile(self, dst_path=None):
+        if not dst_path:
+            dst_path = tde4.getProjectPath()
+        result = tde4.saveProject(dst_path, True)
+        if not bool(result):
```
Comment (BigRoy) [inline openpype/hosts/equalizer/api/host.py:66]: This evaluates the same as just `if not result:` - why is the bool conversion needed?

File: openpype/hosts/equalizer/tests/test_plugin.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    )
+
+
+def get_containers():
+    return get_context_data().get("containers", [])
+
+
+def add_container(container: Container):
+    context_data = get_context_data()
+    containers = get_containers()
```
Comment (BigRoy) [inline openpype/hosts/equalizer/tests/test_plugin.py:78]: This should really be using the output of `get_context_data` since `get_containers` basically queries the same.

File: openpype/lib/log.py [hunk 6]
```diff
@@ -102,6 +102,11 @@ def emit(self, record):
 
         except OSError:
             self.handleError(record)
+        except ValueError:
+            # this is raised when logging during interpreter shutdown
+            # or it real edge cases where logging stream is already closed.
+            # In particular, it happens a lot in 3d equalizer.
+            pass
```
Comment (BigRoy) [inline openpype/lib/log.py:109]: Do you have an example error stack trace for this? `ValueError` seems like an odd case if the logging stream is closed. I feel like instead there might be issues with the call to `self.format()` or whatever.

File: server_addon/equalizer/LICENSE
Comment (BigRoy) [inline server_addon/equalizer/LICENSE:1]: I'm not seeing a license file like these for other hosts - why do we need this included for 3DEqualizer? We're not encapsulating any of their code or plugins, right?

It also mentions no valid year or copyright owner name either.

File: openpype/hosts/equalizer/api/host.py (hunk 4 above)
Comment (antirotor) [reply openpype/hosts/equalizer/api/host.py:66]: because `saveProject()` does not return bool. An yes, it is returnin 1 and 0 on success and python will handle that just as fine, but I like it more explicit.

File: server_addon/equalizer/LICENSE
Comment (antirotor) [reply server_addon/equalizer/LICENSE:1]: This license file is with some other addons like maya or harmony. I used one of them as a template. This will play its role when those will be separated to individual repos. This license file is standard "github generated" Apache 2 license

Comment (BigRoy) [reply server_addon/equalizer/LICENSE:1]: I see, but shouldn't the `[name]` etc. be replaced with e.g. `AYON` or whatever? :)

Also, for consistency - note that the maya addon names it `LICENCE` instead of `LICENSE` - so best to pick one.

File: openpype/lib/log.py (hunk 6 above)
Comment (BigRoy) [reply openpype/lib/log.py:109]: Sounds like we'd be better of investigating more indeed and make this comment a `# TODO` - I'm a bit worried just capturing `ValueError` here silently will instead be hiding other errors in the future that we do want to know about.

File: server_addon/equalizer/LICENSE
Comment (antirotor) [reply server_addon/equalizer/LICENSE:1]: ah, good point :)

File: openpype/hosts/equalizer/api/host.py (hunk 3 above)
Comment (antirotor) [reply openpype/hosts/equalizer/api/host.py:42]: This is just to be sure that host is a singleton. I don't think we enforce it anywhere upstream and I am not sure what exact side-effect there could be if multiple host objects are created, but because of rather fragile way is is using callback timer and storing json encoded context on project notes, I'd rather be sure that it is consistent.

**Comment by seifibrahim32** [comment]
@antirotor I need to communicate with you please.

I have a 3DE4 with older version and I need your help to let me integrate it.
This version supports Python2.7

**Comment by antirotor** [comment]
> @antirotor I need to communicate with you please.
> 
> I have a 3DE4 with older version and I need your help to let me integrate it.
> This version supports Python2.7

For that, you would have to change all py3 syntax like f-strings and type hints to py2.7 syntax. From older 3de API support I can't provide much assistance unfortunately.

**Comment by seifibrahim32** [comment]
@antirotor Frankly I exchanged it at the meantime with .format() as replacement to f-strings but I need to know more details for using hints for python2.7, please.

**Comment by antirotor** [comment]
I guess this can be merged as a base integration and I'll do port to AYON where it can be enhanced further

**Comment by BigRoy** [comment]
> I guess this can be merged as a base integration and I'll do port to AYON where it can be enhanced further

I'd actually move forward and start effort on pushing this into a separate AYON addon - and closing this here. That's at least what I've seen happen with most other open OpenPype contributions? I believe @mkolar closed many of them?

**Comment by antirotor** [comment]
Yeah, for AYON it will become separated, but since this is already done for OP and imho functioning, it would be pitty just not to merge it.

**Comment by antirotor** [comment]
Port to ayon is done here https://github.com/ynput/ayon-equalizer/pull/1
